src/backend/gporca/libgpopt/src/base/CEnfdDistribution.cpp

```diff
@@ -149,6 +150,21 @@ CEnfdDistribution::Epet(CExpressionHandle &exprhdl, CPhysical *popPhysical,
 			return EpetProhibited;
 		}
 
+		if (CDistributionSpec::EdtNonSingleton == m_pds->Edt() &&
+			CDistributionSpec::EdtStrictReplicated == pds->Edt())
```

Comment:
Can you describe meaning of these distributions for the bogus plan provided in the ticket description?

Reply:
As far I understand, we require enforcement for any subtree with Strict Replicated distribution, when Non Singleton distribution is requested. But it seems to be incorrect in the common case. How enforcement is chosen?

Reply:
> But it seems to be incorrect in the common case

Yes and I do that only if CTE producer exists in plan

Reply:
> How enforcement is chosen?

https://github.com/arenadata/gpdb/blob/46f40f274b7ebcc5a8d42076d49c786849931d47/src/backend/gporca/libgpopt/src/search/CJobGroupExpressionOptimization.cpp#L597-L603

Reply:
> Can you describe meaning of these distributions for the bogus plan provided in the ticket description?

this corresponds to Seq Scan on r in our plan

Reply:
> any Broadcast Motion will be

not any, but only under replicated distribution when non-singleton required